capsules/src/net/thread/tlv.rs

```diff
@@ -0,0 +1,1411 @@
+//! only when a variable-length value must be decoded from network byte
+//! order before it can be interpreted correctly. Excluded from this case
+//! are variable-length values that contain data that must later be
+//! decoded by the caller before being interpreted (for example,
+//! sub-TLVs). Such a value is instead returned as a slice of the original
+//! buffer passed to the decode function.
+//!
+//!
+//! Author: Mateo Garcia
+//!         mateog@stanford.edu
+
+
+// TODO: Move the MLE explanation above to the MLE module, when it is created.
+
+
+// NOTES FOR DEBUGGING:
+// - .to_be() may not have been called on values wider than one byte
+// - encode_bytes_be may have been used instead of encode_bytes
+// - decode_bytes_be may have been used instead of decode_bytes
+// - See 4.5.25 Active Operational Dataset TLV and 4.5.26 Pending Operational Dataset TLV
+//    - Are Active and Pending Timestamp TLVs, respectively, required to be sent as well
+//      if either of the dataset tlvs are sent?
+
+
+use core::mem;
+use net::stream::{decode_u8, decode_u16, decode_u32, decode_bytes_be};
+use net::stream::{encode_u8, encode_u16, encode_u32, encode_bytes, encode_bytes_be};
+use net::stream::SResult;
+
+const TL_WIDTH: usize = 2; // Type and length fields of TLV are each one byte.
+const MAX_VALUE_LENGTH: usize = 128; // Assume a TLV value will be no longer than 128 bytes.
+
+/// Type-Length-Value structure.
+pub enum Tlv<'a> {
+    SourceAddress(u16),
+    Mode(u8),
+    Timeout(u32),
+    Challenge([u8; 8]), // Byte string max length 8 bytes.
+    Response([u8; 8]), // Byte string max length 8 bytes.
+    LinkLayerFrameCounter(u32),
+    // LinkQuality,                  // TLV type Not used in Thread
+    // NetworkParameter,             // TLV type Not used in Thread
+    MleFrameCounter(u32),
+    /*
+    TODO: Not required to implement MLE for SED
+    Route64,
+    */
+    Address16(u16),
+    LeaderData {
+        partition_id: u32,
+        weighting: u8,
+        data_version: u8,
+        stable_data_version: u8,
+        leader_router_id: u8,
+    },
+    NetworkData(&'a [u8]),
+    TlvRequest(&'a [u8]),
+    ScanMask(u8),
+    Connectivity {
+        parent_priority: u8,
+        link_quality_3: u8,
+        link_quality_2: u8,
+        link_quality_1: u8,
+        leader_cost: u8,
+        id_sequence: u8,
+        active_routers: u8,
+        sed_buffer_size: Option<u16>,
+        sed_datagram_count: Option<u8>,
+    },
+    LinkMargin(u8),
+    Status(u8),
+    Version(u16),
+    /*
+    TODO: Not required to implement MLE for SED
+    AddressRegistration
+    AddressRegistration
+    Channel
+    PanId
+    ActiveTimestamp
+    PendingTimestamp
+    ThreadDiscovery
+    */
+    ActiveOperationalDataset(&'a [u8]),
+    PendingOperationalDataset(&'a [u8]),
+}
+
+impl<'a> Tlv<'a> {
+    /// Serializes TLV data in `buf` into the format specific to the TLV
+    /// type.
```

Comment:
My understanding from reading the module-level comment is that these `&mut [u8]` buffers should be at least 128 bytes long for the various `encode` methods to always run successfully. Is that corrent?